File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, K):
    heapq.heapify(scoville)
    cnt = 0
    while True:
        first = heapq.heappop(scoville)
        if first >= K:
            break
        if not scoville:  # 만약 한 개만 남아있는 경우
            return -1
        second = heapq.heappop(scoville)
        new_scoville = first + (second * 2)
        heapq.heappush(scoville, new_scoville)
        cnt += 1
    return cnt

File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <algorithm>
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    make_heap(scoville.begin(), scoville.end(), greater<int>());

    while (scoville.size() > 1) {
        int first = scoville.front();
        if (first >= K)
            break;
        pop_heap(scoville.begin(), scoville.end(), greater<int>());
        scoville.pop_back();
        
        int second = scoville.front();
        pop_heap(scoville.begin(), scoville.end(), greater<int>());
        scoville.pop_back();

        scoville.push_back(first + second * 2);
        push_heap(scoville.begin(), scoville.end(), greater<int>());
        
        answer++;
    }
    if (scoville.front() < K) {
        return -1;
    }
    return answer;
}